File: 210-CourseScheduleIi/210-CourseScheduleIi.java
// Last updated: 8/16/2026, 9:51:14 PM
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        // find cycle in directed graph, topological sort - DFS
        // next ligitable node will be pushed to the head of the order -> Stack, LIFO
        Stack<Integer> order = new Stack<>();

        // build graph[i] = adajacent nodes of i / next ready courses
        List<List<Integer>> graph = new ArrayList<>();
        for(int i = 0; i < numCourses; i ++) graph.add(new ArrayList<>());

        for(int[] pair: prerequisites) {
            graph.get(pair[1]).add(pair[0]);
        }
        System.out.println(graph);
        // marked[]: T-processed already, F-not processed so far
        boolean[] marked = new boolean[numCourses];

        // status[]: 0-unknown, 1-visiting, 2-visited
        int[] status = new int[numCourses];

        for(int i = 0; i < numCourses; i ++) {
            if (!marked[i]) {
                if( !dfs(graph, status, i, order) ) return new int[0]; // cycle found, false
                marked[i] = true;
            }
        }

        int[] res = new int[numCourses];
        for(int i = 0; i < numCourses; i ++) res[i] = order.pop();
        return res;
    }

    private boolean dfs(List<List<Integer>> graph, int[] status, int i, Stack<Integer> order) {
        if(status[i] == 1) return false; // cycle found
        if(status[i] == 2) return true; // visited before

        // mark current node-i as visiting
        status[i] = 1;

        // iterate adjacent nodes of i
        for(int j : graph.get(i)) {
            if( !dfs(graph, status, j, order) ) return false; // if node-j found cycle, false
        }

        // mark node-i as visited
        status[i] = 2;
        // add visited node to Stack-order
        order.push(i);
        return true;
    }
}

File: 210-CourseScheduleII/210-CourseScheduleII.py
# Last updated: 8/30/2026, 9:52:41 PM
1class Solution:
2    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
3        # topo sort
4        graph = defaultdict(list)   # node -> [downstream nodes list]
5        indegree = [0] * numCourses
6
7        for a, b in prerequisites:  # edge: b -> a
8            graph[b].append(a)
9            indegree[a] += 1
10
11        queue = deque([c for c in range(numCourses) if indegree[c] == 0])
12        res = []
13
14        while queue:
15            node = queue.popleft()
16            res.append(node)
17
18            for nxt in graph[node]:
19                indegree[nxt] -= 1
20                if indegree[nxt] == 0:
21                    queue.append(nxt)
22        
23        return res if len(res) == numCourses else []